File: bot.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

const { PuppeteerScreenRecorder } = require('puppeteer-screen-recorder');
const cloudinary = require('cloudinary').v2;
const admin = require('firebase-admin');
const { GoogleGenerativeAI } = require("@google/generative-ai");
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));
const path = require('path');
const fs = require('fs');

// Configurations
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(JSON.parse(process.env.FIREBASE_KEY))
  });
}
const db = admin.firestore();

// Settings
const MAX_WAIT_FOR_MEDIA = 45000; 

async function runBot() {
  console.log("🚀 GHOST_ENGINE: V22 - AI + WhatsApp + Profile Tracker Loaded...");
  
  const browser = await puppeteer.launch({ 
    headless: "new", 
    args: ['--no-sandbox', '--disable-setuid-sandbox'] 
  });
  
  const page = await browser.newPage();
  await page.setViewport({ width: 1080, height: 1920 }); 

  try {
    await page.goto('https://www.instagram.com/', { waitUntil: 'networkidle2' });
    if (process.env.INSTA_COOKIES) {
        await page.setCookie(...JSON.parse(process.env.INSTA_COOKIES));
        await page.reload({ waitUntil: 'networkidle2' });
    }

    const targets = ["_anshu_2101", "_cool_butterfly_.6284", "dee_pu3477", "ritu_singh785903"];

    for (const user of targets) {
      console.log(`\n🕵️ Target: @${user}`);
      
      // --- 1. PROFILE TRACKING (Bio & DP) ---
      await trackProfileChanges(page, user);

      // --- 2. HIGHLIGHT DISCOVERY ---
      await page.goto(`https://www.instagram.com/${user}/`, { waitUntil: 'networkidle2' });
      const highlightLinks = await page.evaluate(() => {
          return Array.from(document.querySelectorAll('a[href*="/stories/highlights/"]')).map(a => a.href);
      });

      // --- 3. DEEP SCAN (Stories & Highlights) WITH AUTO-LIKE ---
      // Stories
      await page.goto(`https://www.instagram.com/stories/${user}/`, { waitUntil: 'networkidle2' });
      await captureAndLike(page, user, 'stories_v22');

      // Highlights
      for (const hUrl of highlightLinks) {
          await page.goto(hUrl, { waitUntil: 'networkidle2' });
          await captureAndLike(page, user, 'highlights_v22');
      }
    }

  } catch (error) {
    console.error("❌ CRITICAL ERROR:", error.message);
  } finally {
    await browser.close();
    console.log("⏹️ Bot Mission Finished.");
  }
}

/**
 * ❤️ Auto-Like & Media Capture
 */
async function captureAndLike(page, username, category) {
    for (let i = 0; i < 25; i++) {
        await new Promise(r => setTimeout(r, 6000)); // Media loading wait

        // 🔥 AUTO LIKE LOGIC
        await page.evaluate(() => {
            const likeBtn = document.querySelector('span[role="button"] svg[aria-label="Like"]')?.closest('button');
            if (likeBtn) {
                likeBtn.click();
                console.log("   ❤️ Liked the post!");
            }
        });

        const media = await page.evaluate(() => {
            const video = document.querySelector('video');
            const img = document.querySelector('img[decode="sync"]') || document.querySelector('img[alt*="Story"]');
            if (video && video.readyState >= 3) return { url: video.currentSrc || video.src, type: 'video' };
            if (img && img.complete) return { url: img.src, type: 'image' };
            return null;
        });

        if (media && media.url) {
            const isNew = await safeSync(media.url, username, category, media.type === 'video');
            
            // 🤖 GEMINI AI & WHATSAPP (Only for new items)
            if (isNew) {
                const aiDesc = await analyzeMedia(media.url);
                await updateAIDesc(username, media.url, aiDesc);
                await sendWhatsApp(`🔔 NEW CONTENT: @${username} ne kuch naya dala h!\n\n🤖 AI: ${aiDesc}\n📂 Type: ${media.type}`);
            }
        }

        await page.keyboard.press('ArrowRight');
        await new Promise(r => setTimeout(r, 2000));
        if (!await page.evaluate(() => window.location.href.includes('/stories/'))) break;
    }
}

/**
 * 👤 Profile Tracking (DP & Bio Changes)
 */
async function trackProfileChanges(page, username) {
    const profile = await page.evaluate(() => ({
        bio: document.querySelector('header section div:nth-child(3) span')?.innerText || "",
        dp: document.querySelector('header img')?.src || ""
    }));

    const docRef = db.collection("profile_tracking").doc(username);
    const doc = await docRef.get();

    if (!doc.exists || doc.data().bio !== profile.bio || doc.data().dp !== profile.dp) {
        await docRef.set({ ...profile, updated_at: admin.firestore.FieldValue.serverTimestamp() });
        await sendWhatsApp(`👤 PROFILE UPDATED: @${username} ka Bio ya DP change hua h! Check karo.`);
    }
}

/**
 * 🤖 Gemini AI Image Analysis
 */
async function analyzeMedia(url) {
    try {
        const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });
        const prompt = "Briefly describe what is happening in this Instagram image for a database log.";
        const result = await model.generateContent(prompt); // Note: Simple prompt for speed, can be enhanced
        return result.response.text();
    } catch (e) { return "No description available."; }
}

/**
 * 📱 WhatsApp Notification (CallMeBot)
 */
async function sendWhatsApp(message) {
    const phone = process.env.WHATSAPP_PHONE;
    const apiKey = process.env.WHATSAPP_API_KEY;
    const url = `https://api.callmebot.com/whatsapp.php?phone=${phone}&text=${encodeURIComponent(message)}&apikey=${apiKey}`;
    try { await fetch(url); } catch (e) {}
}

async function safeSync(url, username, category, isVideo) {
    const mediaId = url.split('?')[0].split('/').pop().substring(0, 40); 
    const docId = `V22_${username}_${mediaId}`;
    const docRef = db.collection("archives").doc(docId);
    
    const doc = await docRef.get();
    if (doc.exists) return false;

    const upload = await cloudinary.uploader.upload(url, { 
        folder: `insta_vault_v22/${username}/${category}`, 
        resource_type: isVideo ? "video" : "image"
    });

    await docRef.set({ owner: username, url: upload.secure_url, type: category, time: admin.firestore.FieldValue.serverTimestamp() });
    return true;
}

async function updateAIDesc(username, url, desc) {
    const mediaId = url.split('?')[0].split('/').pop().substring(0, 40);
    const docId = `V22_${username}_${mediaId}`;
    await db.collection("archives").doc(docId).update({ ai_description: desc });
}

runBot();

File: bot.py
import os
import json
import asyncio
import cloudinary
import cloudinary.uploader
import firebase_admin
from firebase_admin import credentials, firestore
import google.generativeai as genai
from playwright.async_api import async_playwright
from playwright_stealth import stealth_async
import requests

# ⚙️ CONFIG
cloudinary.config(
    cloud_name=os.environ.get('CLOUDINARY_NAME'),
    api_key=os.environ.get('CLOUDINARY_API_KEY'),
    api_secret=os.environ.get('CLOUDINARY_API_SECRET')
)
genai.configure(api_key=os.environ.get('GEMINI_API_KEY'))

if not firebase_admin._apps:
    cred_json = json.loads(os.environ.get('FIREBASE_KEY'))
    firebase_admin.initialize_app(credentials.Certificate(cred_json))
db = firestore.client()

captured_media_urls = {}

# --- Gemini AI Analysis ---
async def analyze_with_gemini(image_url):
    try:
        model = genai.GenerativeModel('gemini-1.5-flash-latest')
        prompt = "Identify Area and Describe: [Area] - [Action]"
        response = requests.get(image_url)
        if response.status_code == 200:
            result = model.generate_content([
                prompt,
                {"mime_type": "image/jpeg", "data": response.content}
            ])
            return result.text
    except Exception as e:
        print(f"      ⚠️ AI Error: {e}")
    return "AI analysis unavailable."

# --- Firebase & Cloudinary Sync ---
async def safe_sync(url, username, category, is_video):
    try:
        media_id = url.split('?')[0].split('/')[-1][:45]
        doc_id = f"V28_PY_{username}_{media_id}"
        doc_ref = db.collection("archives").document(doc_id)

        if doc_ref.get().exists:
            return False, None

        print(f"      📤 Uploading to Cloudinary...")
        upload = cloudinary.uploader.upload(
            url, 
            folder=f"insta_vault_v28/{username}/{category}",
            resource_type="video" if is_video else "image"
        )

        doc_ref.set({
            "owner": username,
            "url": upload['secure_url'],
            "type": category,
            "is_video": is_video,
            "time": firestore.SERVER_TIMESTAMP,
            "version": "V28_PLAYWRIGHT_PYTHON"
        })
        return True, doc_id
    except Exception as e:
        print(f"      ❌ Sync Error: {e}")
        return False, None

# --- Network Interception ---
async def handle_request(request):
    url = request.url
    if "instagram.com" in url and ("video" in url or ".mp4" in url or ".jpg" in url):
        if "profile_pic" not in url and "data:image" not in url:
            is_video = "video" in url or ".mp4" in url
            captured_media_urls[url] = {"type": "video" if is_video else "image"}

async def run_bot():
    async with async_playwright() as p:
        print("🚀 GHOST_ENGINE: V28 (Python) - STARTING...")
        browser = await p.chromium.launch(headless=True, args=["--no-sandbox"])
        context = await browser.new_context(viewport={'width': 1080, 'height': 1920})
        page = await context.new_page()
        await stealth_async(page)

        # 🍪 Apply Cookies
        cookies_env = os.environ.get('INSTA_COOKIES')
        if cookies_env:
            print("🍪 Injecting Cookies...")
            await context.add_cookies(json.loads(cookies_env))

        page.on("request", handle_request)
        targets = ["_anshu_2101", "_cool_butterfly_.6284", "dee_pu3477", "ritu_singh785903"]

        for user in targets:
            print(f"\n🕵️ Target Locked: @{user}")
            try:
                # Page load logic with Status check
                response = await page.goto(f"https://www.instagram.com/stories/{user}/")
                
                # --- ERROR LOGGING SYSTEM ---
                if response.status in [403, 429]:
                    print(f"🚨 ERROR: IP BLOCK HAI! (Status: {response.status})")
                    print("💡 Tip: GitHub Actions ki IP block ho gayi hai. Workflows mein macos-latest use karein.")
                    break
                
                await asyncio.sleep(5) # Wait for page content

                if "login" in page.url or await page.query_selector('input[name="username"]'):
                    print("🚨 ERROR: COOKIES EXPIRED/BLOCK HAIN!")
                    print("💡 Tip: Browser se naya 'sessionid' nikaal kar GitHub Secrets update karein.")
                    break

                print(f"✅ Access Granted! Scanning @{user}...")

                for slide in range(10): 
                    if "/stories/" not in page.url: break
                    
                    # SCREEN SMASH: Clear overlays
                    await page.mouse.click(540, 960)
                    await asyncio.sleep(2)

                    if captured_media_urls:
                        latest_url = list(captured_media_urls.keys())[-1]
                        media_data = captured_media_urls[latest_url]
                        
                        is_video = media_data['type'] == 'video'
                        is_new, doc_id = await safe_sync(latest_url, user, "stories", is_video)
                        
                        if is_new and not is_video:
                            desc = await analyze_with_gemini(latest_url)
                            db.collection("archives").document(doc_id).update({"ai_report": desc})
                    
                    await page.keyboard.press("ArrowRight")
                    await asyncio.sleep(3)

            except Exception as e:
                print(f"❌ System Error scanning @{user}: {e}")

        await browser.close()
        print("⏹️ Mission Finished.")

if __name__ == "__main__":
    asyncio.run(run_bot())
